Add v3 upgrade guide (#85)
humanmade/altis-documentation: other-docs/guides/upgrading/v3.md

# Upgrading to v3

To upgrade to Altis v3, edit your `composer.json` and change the version constraint for `altis/altis` and any local environment modules to `^3.0.0`:

```json
{
	"require": {
		"altis/altis": "^3.0.0"
	},
	"require-dev": {
		"altis/local-chassis": "^3.0.0",
		"altis/local-server": "^3.0.0"
	}
}
```

Next remove the `vendor` directory by running `rm -rf vendor` or on Windows `rmdir vendor`. You could also delete the directory using your code editor, finder or explorer.

*Note:* due to an issue in how Composer handles installation of `composer-plugin` packages the above step is required to ensure the new version of the required packages are used to manage the process.

Next run `composer update` to complete the upgrade. You should commit the updated `composer.json` and `composer.lock` files.

## Breaking Changes

### Media

> Intermediate image size files are no longer created but handled entirely by Tachyon if `modules.media.smart-media` is enabled in your config, this is on by default. This is only a breaking change if you have code that relies on fetching those files (not the original or full size image) directly by their path.

Reviewer: ```suggestion
Files for intermediate image sizes are no longer created but handled entirely by Tachyon if `modules.media.smart-media` is enabled in your config, which is the default. This is only a breaking change if you have code that relies on fetching those files (not the original or full size image) directly by their path.
```

### Block Editor

> The updated Block Editor in WordPress 5.3 has many updates to the underlying javascript packages such as React itself and features of the `@wordpress/data` package. If you make use of helpers like `wp.data.withDispatch` you should test this update thoroughly.

Reviewer: ```suggestion
The updated Block Editor in WordPress 5.3 has many updates to the bunlded JavaScript packages such as React , as well as features in the `@wordpress/data` package. If you make use of helpers like `wp.data.withDispatch` you should test this update thoroughly.
```

> The updated Block Editor in WordPress 5.3 has many updates to the bunlded JavaScript packages such as React , as well as features in the `@wordpress/data` package. If you make use of helpers like `wp.data.withDispatch` you should test this update thoroughly.

Reviewer: ```suggestion
The updated Block Editor in WordPress 5.3 has many updates to the bundled JavaScript packages such as React, as well as features in the `@wordpress/data` package. If you make use of helpers like `wp.data.withDispatch` you should test this update thoroughly.
```

## Headline Features

### WordPress 5.3 "Kirk"

The latest CMS Module brings with it WordPress version 5.3. Some of the highlights are:

> - Large image files are stored separately as the original and a web optimised full size version is generated for use on the website to improve performance

Reviewer: ```suggestion
- For super-large images, image resizing now uses a scaled-down version of the original image, significantly improving frontend performance
```

- The Block Editor has many accessibility improvements and enhanced layout features
- Automatic image rotation
- Administrator email verification
- PHP 7.4 Support

You can [find the full WordPress 5.3 release notes here](https://wordpress.org/news/2019/11/kirk/).

### Zero Configuration PHPUnit Testing

Writing and runnning PHPUnit tests for your application is now a simple process. Add your test classes to a directory called `tests` in your project root directory and run `composer dev-tools phpunit`.

> Tests are run on the local environment whether it is [Local Server](docs://local-server) or [Local Chassis](docs://local-chassis) and as such provide full integration testing capabilities to match production with S3, ElasticSearch, caching and Tachyon.

Reviewer: I think we can be a little less verbose:

```suggestion
Tests are run in your local environment, and provide full integration testing capabilities to match production, including S3, ElasticSearch, caching and Tachyon.
```

This feature provides flexibility for more complex use cases, check out the [unit testing documentation and examples here](docs://dev-tools/testing-with-phpunit.md).

### Browser Security

> Good security is a growing concern online and so the [browser security](docs://security/browser.md) feature provides a configurable solution for securing static assets and cross origin behaviour. Supported features include:

Reviewer: ```suggestion
Good security is a growing concern online, and the new [browser security](docs://security/browser.md) feature set provides a configurable solution for securing static assets and cross origin behaviour. Supported features include:
```

> - Automatic Sub Resource Integrity hashes for 1st party scripts with 3rd party support via filters

Reviewer: ```suggestion
- Automatic [Subresource Integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) for first-party assets, and a framework for third-party assets
```

> - Configurable `XSS-Protection` and `X-Frame-Options` headers

Reviewer: ```suggestion
- Automatic `X-XSS-Protection` and `X-Frame-Options` headers by default
```

- Content Security Policy support, with smart defaults and highly configurable

## Other Developer Features and APIs

### Core Module Loading (Core Module)

> Modules were previously loaded via composer's autoload functionality however this approach was less than ideal as the entrypoint files contained side-effects and required a `function_exists()` check at the start of the file to prevent certain composer commands breaking.

Reviewer: ```suggestion
Modules were previously loaded via Composer's autoload functionality, however this approach was limiting, as the entrypoint files contained side-effects and required a `function_exists()` check at the start of the file to prevent non-Altis CLI commands from breaking.
```

The [advice for writing custom modules](docs://getting-started/custom-modules.md) has now been updated to reflect this but any existing custom modules will remain backwards compatible.

### Developer Tools Additions (Developer Tools module)

Query Monitor integration is now vastly improved with clickable traces for both PHP function calls and database queries that open in your editor.

In addition X-Ray data and the flamegraph are now available as panels in Query Monitor.

A [default Travis CI config](docs://dev-tools/continuous-integration.md) will be created if one does not already exist that will run the zero config PHPUnit tests.

### Local Chassis improvements (Local Chassis module)

Along with a `composer chassis exec` command for running single commands directly on the VM [you can now extend the default chassis configuration from your `composer.json`](docs://local-chassis/) like so:

```json
{
	"extra": {
		"altis": {
			"modules": {
				"local-chassis": {
					"hosts": [
						"altis.local",
						"subsite.altis.local"
					],
					"extensions": [
						"xdebug"
					]
				}
			}
		}
	}
}
```

> ### Local Server improvements (Local Server module)

Reviewer: I'd merge these local env sections.

> As with the Local Chassis module there is now a `composer local-server exec` command for running arbitrary commands on the PHP docker container.

Reviewer: Sorry, was imprecise before; let's combine this part with the `chassis exec` part, and just have something like:

> Both Local Server and Local Chassis now support an `exec` command for running arbitrary commands in the development environment. `composer local-chassis exec` runs a command inside the virtual machine, while `composer local-server exec` runs a command inside the PHP Docker container

Author: ah I see

> XDebug is also provided behind a feature flag, run `composer local-server start --xdebug` to start the service with XDebug enabled. [Documentation for using XDebug with popular editors can be found here](docs://local-server/php-debugging.md).

Reviewer: ```suggestion
For convenience, Local Server supports XDebug with a feature flag. Run `composer local-server start --xdebug` to start the server with XDebug enabled. [Documentation for using XDebug with popular editors can be found here](docs://local-server/php-debugging.md).
```

### Security improvements (Security module)

Passwords are now encrypted using BCrypt for enhanced security. In addition XML-RPC can now be switched off via configuration.

> Requiring login now respects individual sites "public" setting. If unchecked then users must be logged in to view the site.

Reviewer: ```suggestion
The Require Login feature now respects the "public" setting on each site. If a site is not public, users must be logged in to access it.
```

### Cloud improvements (Cloud module)

Page caching is now made more performant by ignoring common query string parameters and caching redirects. It is also highly configurable and supports the `Cloudfront-Viewer-Country` header out of the box.

[Check the page caching documentation for more information](docs://cloud/page-caching.md).

> The dashboard has many enhancements including better performance and the ability to pin apps.

Reviewer: ```suggestion
The [Altis Dashboard](https://dashboard.altis-dxp.com/) has many enhancements including better performance and the ability to pin apps.
```
